File: 1221_balanced_string_split/balanced_string_split.go
/*
 * @Date: 2021-09-07 16:57:01
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-09-07 17:10:59
 */

package main

func balancedStringSplit(s string) (ans int) {
	d := 0
	for _, ch := range s {
		if ch == 'L' {
			d++
		} else {
			d--
		}
		if d == 0 {
			ans++
		}
	}
	return
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	{
		s := "RLRRLLRLRL"
		ans := 4
		assert(balancedStringSplit(s) == ans)
	}
	{
		s := "RLLLLRRRLR"
		ans := 3
		assert(balancedStringSplit(s) == ans)
	}
	{
		s := "LLLLRRRR"
		ans := 1
		assert(balancedStringSplit(s) == ans)
	}
	{
		s := "RLRRRLLRLL"
		ans := 2
		assert(balancedStringSplit(s) == ans)
	}
}

File: 1221_balanced_string_split/balanced_string_split.cpp
/*
 * @Date: 2021-09-07 16:56:58
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-09-07 17:02:46
 */

#include <cassert>
#include <string>

using namespace std;

class Solution {
 public:
  int balancedStringSplit(string s) {
    int ans = 0, d = 0;
    for (char ch : s) {
      ch == 'L' ? ++d : --d;
      if (d == 0) ++ans;
    }
    return ans;
  }
};

int main() {
  {
    string s = "RLRRLLRLRL";
    int ans = 4;
    assert(Solution().balancedStringSplit(s) == ans);
  }
  {
    string s = "RLLLLRRRLR";
    int ans = 3;
    assert(Solution().balancedStringSplit(s) == ans);
  }
  {
    string s = "LLLLRRRR";
    int ans = 1;
    assert(Solution().balancedStringSplit(s) == ans);
  }
  {
    string s = "RLRRRLLRLL";
    int ans = 2;
    assert(Solution().balancedStringSplit(s) == ans);
  }
}